feat: KONFLUX-3383 Add initial test selection framework (#1254)
konflux-ci/e2e-tests: magefiles/rulesengine/readme.md

# Mage Engine

> Mage Engine is a rules engine framework with the intent to take the existing businesss 
> logic that dictates configuring the environment and test exection we have 
> accumlated over time in our magefile and organize them better.

Reviewer: ```suggestion
Mage Engine is a rules engine framework with the intent to take the existing business 
logic that dictates configuring the environment and test execution we have 
accumulated over time in our magefile and organize them better.
```

Author: Done

The core of what drives the rules will still be the functions we write that check for 
specific conditions to be met and actions to take on said conditions. The framework 
just allows us to take those higher order functions and compose them together into
more descriptive rules.

There is NO INTENT to be a full fledged rules engine to be able to evaluate very complex expressions!

## Architecture

> The following Architecure of MageEngine. 

Reviewer: ```suggestion
The following Architecture of MageEngine. 
```

Author: Done

### Rule

The core of the engine is a `Rule`. A Rule describes and executes the business logic. It is composed 
of a `Conditional` which implements the checking of data that should be evaluated to `true`. 

> Once a condtional of a `Rule` evaulates true, the `Rule` can take an `Action`. The action implement 

Reviewer: ```suggestion
Once a conditional of a `Rule` evaluates true, the `Rule` can take an `Action`. The action implements 
```

the execution of the task the rule should take.

A `Rule` implements the `IRule` interface: 
 * To evaluate the registered condition the engine calls `Eval()` on the rule.
 * To take action when evaluation is true the engine calls `Apply()` on the rule.
 * To simulate an action on a rule the engine can call `DryRun()` (DryRun needs to be set on the `RuleCtx` 
   for the framework to make this call.)

### Rule Context
a `RuleCtx` is the context object insert data into and gets passed around so that rules can evaluate and take action.
In our use case we have very specific key pieces of data that triggers our business logic executes on so 

> this isn't generalized at the moment to accomodate more generalized use cases. 

Reviewer: ```suggestion
this isn't generalized at the moment to accommodate more generalized use cases. 
```

Author: Done

### Conditionals

`Conditional` is an interface that is meant to perform a check of the passed in data. A conditional can 
be:
 * a full fledged object/struct that implements the `Check()` function 

>  * any anonymous or higher order function as long as it is regisered in the Rule as a `ConditionalFunc`

Reviewer: ```suggestion
 * any anonymous or higher order function as long as it is registered in the Rule as a `ConditionalFunc`
```

Author: Done

#### Filters

> To make constructing `Rules` with some complex conditionals easier the framework implementated a couple of filters which implment

Reviewer: ```suggestion
To make constructing `Rules` with some complex conditionals easier the framework implements a couple of filters which implement
```

Author: Done

the `Conditional` interface:

 * All: `Conditional|ConditionalFunc` registered will evaluate that all evaluated to `true`
 * Any: `Conditional|ConditionalFunc` registered will evaluate that one evaluated to `true`
 * None: `Conditional|ConditionalFunc` registered will evaluate that all have NOT evaluated to `true`. The absensce of `true`

### Actions

> `Action` is an interface that is meant to execute on the passed in data. Like a condtional, an action can 

Reviewer: ```suggestion
`Action` is an interface that is meant to execute on the passed in data. Like a conditional, an action can 
```

Author: Done

be:
 * a full fledged object/struct that implements the `Execute()` function
 * any anonymous or higher order function as long as it is regisered in the Rule as a `ActionFunc`

### RuleCatalog

This is really a collection/slice of Rules. You create a catalog and register the catalog with the engine.
As of now most of the Business Logic really is tied by repo. So catalogs are created by repo/domain

### Engine 

It is a map of a map. The idea was that each repo would create a catalog of Rules. Those Catalogs could be registered 

> as map to a `category`. For exmaple, using the example of test execution, a category could be `tests`. 

Reviewer: ```suggestion
as map to a `category`. For example, using the example of test execution, a category could be `tests`. 
```

Author: Done

Then under `tests` we regsiter a map, where a key is the repo/domain, i.e. `e2e-repo` and its test catalog is assigned to it.  

### RuleChain

A `RuleChain` is more of a concept than it is an actual type. A rulechain is a `Rule` type but it also implements
the `Conditional` interface. So what does that mean? A rule can now be composed of other rules by registering them 
as conditionals.

Since Rules are conditionals as well. The evaluation process is a little different. When a RuleChain evaluates a rule
it will run the `Eval()` and `Apply()|DryRun()`. Both those calls have to execute without fail for the Rule to
evaluate true. 

This will help reuse existing rules to create broader flows.

 ## Example of constructing Rules/RulesCatalogs/RuleChains

 ### Creating a basic rule using anonymous functions and its rule catalog

 In this example, we create a rule for infra-deployments test execution using anonymous condition and action functions. Some 

>  conditions and actions maybe so simple that it is easier to implement as anonymous functions. I then define this Rule within

Reviewer: ```suggestion
 conditions and actions may be so simple that it is easier to implement as anonymous functions. I then define this Rule within
```

Author: Done

 a Rule Catalog. 

>  The logic is simple WHEN the RepoName is infra-deployments, THEN execute ginkgo with the specific lable filter.

Reviewer: ```suggestion
 The logic is simple WHEN the RepoName is infra-deployments, THEN execute ginkgo with the specific label filter.
```

Author: Done

		Condtion:    rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
			if rctx.RepoName == "infra-deployments" {
				return true
			}
			return false
		}),
		Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
			rctx.LabelFilter = "e2e-demo,rhtap-demo,spi-suite,remote-secret,integration-service,ec,build-templates,multi-platform"
			return ExecuteTestAction(rctx)
		})}},

} 
 
 ```

 ### Registering a rule catalog to the engine

 Here I've created a catagory called tests within the MageEngine and I've assigned infra-deployments catalog to the infra-deployments key
 under the tests category

 ```go
 var MageEngine = rulesengine.RuleEngine{
	"tests": {
		"infra-deployments": testselection.InfraDeploymentsTestRulesCatalog,
	},
}
 ```

 ### Executing the MageEngine within a mage file

 This will cause the engine to load and iterate over the rules of the category test and apply any that had their conditions match.

 ```go

    rctx := rulesengine.NewRuleCtx()
	rctx.DryRun = true

	err = engine.MageEngine.RunRulesOfCategory("tests", rctx)
 ```

 ### Creating a complex rule using a mixture of conditional filters

>  In this example, the release-catalog-repo has a slightly more complex set of rules where depending on certain conditons

Reviewer: ```suggestion
 In this example, the release-catalog-repo has a slightly more complex set of rules where depending on certain conditions
```

Author: Done

# Mage Engine

Mage Engine is a rules engine framework with the intent to take the existing business 
logic that dictates configuring the environment and test execution we have 
accumulated over time in our magefile and organize them better.

The core of what drives the rules will still be the functions we write that check for 
specific conditions to be met and actions to take on said conditions. The framework 
just allows us to take those higher order functions and compose them together into
more descriptive rules.

There is NO INTENT to be a full fledged rules engine to be able to evaluate very complex expressions
or assign priority to rules if multiple rules evaluate to true. As of right now it wasn't obvious
if we had such complex requirements based on the data we use.

## Architecture

The following Architecture of MageEngine. 

### Rule

The core of the engine is a `Rule`. A Rule describes and executes the business logic. It is composed 
of a `Conditional` which implements the checking of data that should be evaluated to `true`. 

> Once a conditional of a `Rule` evaulates true, the `Rule` can take an `Action`. The action implements 

Reviewer 2: ```suggestion
Once a conditional of a `Rule` evaluates true, the `Rule` can take an `Action`. The action implements 
```

Author: Done

this isn't meant for generalized use cases.

That said, the `RuleCtx` implements a `AddRuleData(key string, obj any)` and `GetRuleData(key string)`
functions that would allow a developer to store other types of data. It is important to note that
the writer of a conditional or action that depends on data stored in this manner will have to make 
sure they type cast the returned object appropriately in their implementation.

### Conditionals

`Conditional` is an interface that is meant to perform a check of the passed in data. A conditional can 
be:
 * a full fledged object/struct that implements the `Check()` function 
 * any anonymous or higher order function as long as it is registered in the Rule as a `ConditionalFunc`

#### Filters

To make constructing `Rules` with some complex conditionals easier the framework implements a couple of filters which implement the `Conditional` interface:

 * All: `Conditional|ConditionalFunc` registered will evaluate that all evaluated to `true`
 * Any: `Conditional|ConditionalFunc` registered will evaluate that one evaluated to `true`

>  * None: `Conditional|ConditionalFunc` registered will evaluate that all have NOT evaluated to `true`. The absensce of `true`

Reviewer 2: ```suggestion
 * None: `Conditional|ConditionalFunc` registered will evaluate that all have NOT evaluated to `true`. The absence of `true`
```

Author: Done

### Actions

`Action` is an interface that is meant to execute on the passed in data. Like a conditional, an action can 
be:
 * a full fledged object/struct that implements the `Execute()` function

>  * any anonymous or higher order function as long as it is regisered in the Rule as a `ActionFunc`

Reviewer 2: ```suggestion
 * any anonymous or higher order function as long as it is registered in the Rule as a `ActionFunc`
```

Author: Done

### RuleCatalog

This is really a collection/slice of Rules. You create a catalog and register the catalog with the engine.
As of now most of the Business Logic really is tied by repo. So catalogs are created by repo/domain

### Engine 

> It is a map of a map. The idea was that each repo would create a catalog of Rules. Those Catalogs could be registered 

Reviewer 2: ```suggestion
It is a map of a map. The idea was that each repo would create a catalog of Rules. Those catalogs could be registered 
```

Author: Done

Author: Done

as conditionals.

Since Rules are conditionals as well. The evaluation process is a little different. When a RuleChain evaluates a rule
it will run the `Eval()` and `Apply()|DryRun()`. Both those calls have to execute without fail for the Rule to
evaluate true. 

This will help reuse existing rules to create broader flows.

 ## Example of constructing Rules/RulesCatalogs/RuleChains

 ### Creating a basic rule using anonymous functions and its rule catalog

 In this example, we create a rule for infra-deployments test execution using anonymous condition and action functions. Some 
 conditions and actions may be so simple that it is easier to implement as anonymous functions. I then define this Rule within
 a Rule Catalog. 

 The logic in this rule is straightforward: WHEN the RepoName is infra-deployments, THEN execute ginkgo with the specific label filter.

 ```go
 // Example rule for infra-deployments using anonymous functions embedded in the rule
var InfraDeploymentsTestRulesCatalog = rulesengine.RuleCatalog{
	rulesengine.Rule{Name: "Infra Deployments Default Test Execution",
		Description: "Run the default test suites which include the demo and components suites.",

> 		Condtion:    rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {

Reviewer 2: ```suggestion
		Condition:    rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
```

Author: Done

 This will cause the engine to load and iterate over the rules of the category test and apply any that had their conditions match.

 ```go

    rctx := rulesengine.NewRuleCtx()
	rctx.DryRun = true

	err = engine.MageEngine.RunRulesOfCategory("tests", rctx)
 ```

 ### Creating a complex rule using a mixture of conditional filters

 In this example, the release-catalog-repo has a slightly more complex set of rules where depending on certain conditions
 met by the CI job the test should execute with a different set of ginkgo label filters. So you will see a mixture of the 
 Any/All/None conditional filters

 The first rule's logic is: WHEN EITHER RepoName equals release-service-catalog AND NOT pr paired OR RepoName equals release-service-catalog
 AND is a rehearse job AND NOT pr paired THEN run ginkgo with release-pipelines label filter

 The second rule's logic is: WHEN RepoName equals release-service-catalog AND pr is paired AND NOT a rehearse job THEN run ginkgo with 
 release-pipelines && !fbc-tests label filters

 ```go

 rulesengine.Rule{Name: "Release Catalog Test Execution",
		Description: "Runs release catalog tests on release-service-catalog repo on pull/rehearsal jobs.",

> 		Condtion:    rulesengine.Any{rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.None{rulesengine.ConditionFunc(isPaired)}}, rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.ConditionFunc(isRehearse)}, rulesengine.None{rulesengine.ConditionFunc(isPaired)}},

Reviewer 2: ```suggestion
		Condition:    rulesengine.Any{rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.None{rulesengine.ConditionFunc(isPaired)}}, rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.ConditionFunc(isRehearse)}, rulesengine.None{rulesengine.ConditionFunc(isPaired)}},
```

Reviewer 3: From the logic before - if rulesengine.ConditionFunc(isRehearse) is true, no test need to be run.

Author: Done

> 		Condtion:    rulesengine.Any{rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.None{rulesengine.ConditionFunc(isPaired)}}, rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.ConditionFunc(isRehearse)}, rulesengine.None{rulesengine.ConditionFunc(isPaired)}},

Reviewer 3: From the logic before - if rulesengine.ConditionFunc(isRehearse) is true, no test need to be run. Need to move "rulesengine.ConditionFunc(isRehearse)" to rulesengine.None{..}.

Author: Yes, I think you're right in the case of putting the settings in place because the setting up of specific values is a little more complex. I was thinking we could create those as separate rules. These example rules I was focused more on on test execution so when your team uses `release-pipelines` vs `release-pipeliens && !fbc-tests` label filter

That said, I reviewed the logic in the example and in the actual mage code and I think I overthought the conditional logic to try to catch for corner cases that may never happen in CI execution. I simplified the logic and it should follow what we have in the mage file. Let me know if it is clearer

		Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
			rctx.LabelFilter = "release-pipelines"
			return ExecuteTestAction(rctx)
		})}},
rulesengine.Rule{Name: "Release Catalog PR paired Test Execution",
		Description: "Runs release catalog tests except for the fbc tests on release-service-catalog repo when PR paired and not a rehearsal job.",

> 		Condtion:    rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.ConditionFunc(isPaired), rulesengine.None{rulesengine.ConditionFunc(isRehearse)}},

Reviewer 2: ```suggestion
		Condition:    rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.ConditionFunc(isPaired), rulesengine.None{rulesengine.ConditionFunc(isRehearse)}},
```

Author: Done

> 		Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {

Reviewer 3: Need to add this [part](https://github.com/konflux-ci/e2e-tests/blob/main/magefiles/magefile.go#L503C1-L510C66) in the Actions later. It will be registered as action, right?

Author: In theory, Yes, that is how it would work! But talking with Pavel setting those variables in this Rule is too late since bootstrapping depends on these variables.
